sem: fix infinite loop in type identifier lookup

saem opened the pull request:
## Summary

Fixed an infinite loop in type identifier lookup, when parameter names
collided with parameter type declarations.

Fixes https://github.com/nim-works/nimskull/issues/1151

## Details

`semtypes.semTypeIdent`  had an infinite loop when searching for
potentially ambiguous symbols. This meant code with ambiguities between,
for instance, a parameter and a type name resulted in an infinite loop.

In addition to the change a test has been added to avoid the
regression.

saem committed:
sem: fix infinite loop in type identifier lookup

Summary
=======

Fixed an infinite loop in type identifier lookup, when parameter names
collided with parameter type declarations.

fixes: https://github.com/nim-works/nimskull/issues/1151

Details
=====

`semtypes.semTypeIdent` had an infinite loop when searching for
potentially ambiguous symbols. This meant code with ambiguities between,
for instance, a parameter and a type name resulted in an infinite loop.

In addition to the change a test has been added to avoid the regression.

zerbina reviewed (commented):
My understanding of the problem, in the form of a test description, is that:
```
Ensure types are preferred in type contexts when the name is
overloaded with parameter and/or generic parameter names.
```
This differs quite a bit from the current test description. I'm not sure I understand the "repeat the same type" part, could you elaborate on what you mean?

saem commented:
> This differs quite a bit from the current test description.

I struggled with the description, so I'm happy to update it.

> I'm not sure I understand the "repeat the same type" part, could you elaborate on what you mean?

Two things: 1) that two identical type declarations shouldn't be a problem, and 2) it's in 'response' to the comment in `semTypeIdent`, which is `# this implements the wanted ``var v: V, x: V`` feature`. Admittedly, wrt to 1, it's barely a test, and wrt to 2, it's not really a feature.

zerbina commented:
> Two things: 1) that two identical type declarations shouldn't be a problem, and 2) it's in 'response' to the comment in `semTypeIdent`, which is `` # this implements the wanted ``var v: V, x: V`` feature ``. Admittedly, wrt to 1, it's barely a test, and wrt to 2, it's not really a feature.

Thanks. Personally, I'd leave both the "repeat the same part" and `proc f(v: V, x: V)` part out.

---

As an aside (i.e., no change expected in the context of this PR), the `this implements the wanted...` comment in `semTypeIdent` should probably be removed/reworded, as it doesn't apply at all. `var v: V, x: V` isn't valid (and even if it were, `semTypeIdent` has nothing to do with it), and `proc f(v: V, x: V)` would do without the symbol resolution, as `v` and `V` are different identifiers.

saem committed:
remove misleading comment

saem committed:
test description fixup and add generic param test

saem commented:
Updated:

- removed the misleading comment
- added a generic parameter related test
  - noted that generic parameter collisions should probably lead to an ambiguity error

zerbina commented:
/merge
